### DISCUSSION TITLE
Make Linux dynamic linkage a first class citizen

### DISCUSSION THREAD

**Discussion by dweckmann** [microsoft/vcpkg#19127] [Ideas]
****

**Is your feature request related to a problem? Please describe.**
Currently, and on the contrary to Windows platform, the official triplet on Linux is the one with static linkage. There are many problems with that:
1. The CI only test static linkage, which means the quality is not always paired with static build.
2. Static linking make linking big libraries like Qt very slow. Really, when debugging, this is painful
3. License problem: LGPL forbid static linking. Using static linking means your code will be GPL licensed. This is not acceptable for many of us.

**Proposed solution**
- Add an official dynamic linkage triplet for linux, make them default, like on windows 
- Add a CI step to test it

**Comment by JackBoosY** [comment]
Hmm... @BillyONeal what do you think about?

**Comment by BillyONeal** [comment]
We choose static linking on the POSIX-ish platforms because dynamic linking generally forces you to use the system-provided libraries (e.g. those acquired through apt). It doesn't matter if you want to use the vcpkg version of boost in general, for example, because all the operating system's default load library paths are going to look in places owned by apt or dnf or similar.

Windows is different because on Windows your app can use version 1 of a shared library, and someone else can use version 2 of that shared library, and as long as those versions have different DLL names, a program can use both versions, since all symbols from that library are namespaced into the DLL in question. In contrast, in POSIX world, all the names go into the shared process symbol table, so trying to load multiple versions of a library into the same process will usually result in memory corruption.

(This is part of the problem domain and affects all non-system-wide package management solutions, so our ability to influence this is limited)

**Comment by dweckmann** [comment]
Yes, you are indeed right.

However, checking, at least, that the package build correctly on the CI would still increase the quality of the dynamic linkage packages. I guess the package must build out-of-the-box, in all case, even if static linking is the default (which I can live with it).

In the posix world, you have some mean to control where the dynamic loader finds their libraries (most notably by setting LD_LIBRARY_PATH to VCPKG 'installed' lib path). Another option, but this is harder, would be to use `rpath` at linking stage to ensure to load the dependencies of a library with a relative path from themself, so not the system one...

Agree, this is not easy, but other package management (Conan for stance) are able to (or they claim to) somewhat ensure that.

**Comment by BillyONeal** [comment]
>Agree, this is not easy, but other package management (Conan for stance)

While it's true that other package managers have decided to make dynamic linking on the POSIX platforms a more first class citizen, we believe that leads customer confusion; hence our strong recommendation of static linking on those platforms, and that triplet's community status.

We do have x64-linux-dynamic, and we do accept PRs that improve the experience for that triplet. However, even for folks who only use vcpkg locally, we believe it is too easy to accidentally get runtime libraries from the system at runtime resulting in "impossible" memory corruption, and we don't want to handle that support burden right now. For anything targeting wider distribution, those issues are much worse, since it's it might work due to `rpath` or similar on the dev box and then break when copied to an end user.

Despite the above, if some work is contributed to improve the experience for dynamic on those platforms, such as copying the right `.so`s in integrated build systems on the POSIX-ish platforms, that might be a reason to take that triplet out of community status. I think changing the default will remain off the table.

**Comment by dweckmann** [comment]
Yes, we already know the implication of this linkage type. However, it's not a matter of taste for us. The licence and the GPL contamination is a no go, as, even if most of our work are free and open (take a look at https://github.com/IRCAD-IHU/sight), some external part, are proprietary (just because it uses confidencial hardware, software, stuff that waits for a publication in a medical journal, ..) and we have no way to change that. Using for example static Qt (or any LGPL libraries) would mean we cannot go further than private and confidential prototype. No way to use our software in a OP room.

If your strong recommendations stays on static linkage, that's fair and understandable. We only beg for a CI support, to be sure that things compiles and that a submission doesn't break everything. We are in the process to evaluate VCPKG as a replacement for conan, and that is a real concern for us.

Anyway, thanks for listening :)

**Comment by BillyONeal** [comment]
>We do have x64-linux-dynamic,

Correction, we don't have that 😳

**Comment by Osyotr** [comment]
UPD: `x64-linux-dynamic` is now included in vcpkg! https://github.com/microsoft/vcpkg/pull/23035

~I've been able to build some ports with this overlay triplet:~
```cmake
set(VCPKG_TARGET_ARCHITECTURE x64)
set(VCPKG_CRT_LINKAGE dynamic)
set(VCPKG_LIBRARY_LINKAGE dynamic)

set(VCPKG_CMAKE_SYSTEM_NAME Linux)

# set rpath for libraries
if(PORT MATCHES "fribidi|harfbuzz|pkgconf")
    # Meson, nuff said
    set(pretty_ORIGIN_name "\$ORIGIN")
    set(VCPKG_LINKER_FLAGS "-Wl,-rpath,${pretty_ORIGIN_name}:${pretty_ORIGIN_name}/../lib")
elseif(PORT MATCHES "python2")
    # setting rpath here breaks build and I have no idea how to fix it.
    # But since python2 has no dependencies, not setting rpath is ok
else()
    set(VCPKG_LINKER_FLAGS "-Wl,-rpath,'$ORIGIN':'$ORIGIN/../lib'")
endif()
```

**Reply by MrBolt2005** [reply]
Why the specific treatment to packages such as fribidi, harfbuzz and pkgconf? I didn't get it _yet_, _really_...

**Reply by Osyotr** [reply]
@MrBolt2005 because they use meson, which is an "extraordinary 
special kind of flower". There is a bug somewhere when you pass strings with quotes and I really don't want to dig in it.

**Comment by abique** [comment]
@BillyONeal I think there are new options today with `dlmopen` and `RTLD_DEEPBIND`. But I'm not a shared object expert.

**Comment by JackBoosY** [comment]
To add dynamic UNIX support, we must first deal with the default behavior of the system to find dynamic libraries.

**Reply by dg0yt** [reply]
What default behaviour does this refer to? Running binaries from the `tools` directory?

**Reply by JackBoosY** [reply]
According to my pool UNIX knowledge, since the system in UNIX searches for dynamic libraries based on the contents of certain configurations (such as ld.so.conf), we must add _installed/TRIPLET/lib_ and _installed/TRIPLET/lib_ (according to the configuration) to it.

**Reply by Neumann-A** [reply]
https://github.com/microsoft/vcpkg/discussions/19127#discussioncomment-1049627
Just adjust the rpath setting and teach vcpkg_copy_tools and vcpkg_copy_tool_dependencies to also copy .so

**Reply by BillyONeal** [reply]
@Neumann-A The problem isn't vcpkg tools. The problem is the resulting libraries, which are almost never used from the location where they are built. You would need something to patch the rules, e.g. https://github.com/microsoft/vcpkg/pull/15134

**Reply by Neumann-A** [reply]
>  The problem is the resulting libraries, which are almost never used from the location where they are built.

Whats the problem with having a CMAKE_TOOLCHAIN_FILE (or the equivalent VCPKG_.... in the triplet) with the linker flag `-rpath $ORIGIN` ? (no patchelf needed)
If vcpkg does its job correctly every library/executable gets build with that setting. Then it is just a matter of copying the .so like .dlls on windows (or am I missing something?)

Of course consumers of the libraries should then also follow this "windows" way of shared libs for deployment.

**Reply by BillyONeal** [reply]
That might also work. The point is that the resulting catalog has to be useful for something, which isn't currently true.

**Reply by dg0yt** [reply]
Not sure what is expected to be useful ot not. 
Without dynamic linkage, which worked for my app+Qt+GDAL for macOS and Android in the past, vcpkg is not going to be useful.

**Reply by BillyONeal** [reply]
> Not sure what is expected to be useful ot not.

Currently, if you build with a dynamic triplet on Linux, your program will fail to start, because the place where vcpkg puts the resulting .sos won't be found by the operating system loader. This 'rpath' feature might be able to fix that, either by using `patchelf` (as in #15134) or by passing linker flags (as @Neumann-A indicates above). Users needing to manually change environment variables to point into vcpkg guts to run a program doesn't seem like a useful outcome.

**Reply by dg0yt** [reply]
Okay. So consistently setting linker flags for the major build systems, then testing tools installed by ports, and eventually using the "catalog" e.g. in manifest mode, to run tests and to debug user apps etc. is the acceptance tests for a community triplet?

**Reply by Osyotr** [reply]
So, what options do we have? My "solution" above does not scale much, because in manifest mode introduction of new lib with, say, meson as buildsystem would require modifying triplet file thus invalidating binary cache. Setting rpath for every port is also _inconvenient_.
I personally think that Linux shared object location resolution is broken beyond repairing and vcpkg should pass `"-Wl,-rpath,'$ORIGIN'` to linker to mimic Windows behavior entirely. I'm not sure where we need to make changes, though. My guess is `vcpkg_configure_*.cmake` scripts.

**Reply by BillyONeal** [reply]
That seems like it would be reasonable behavior, yes. Unfortunately I am also not positive where such edits would need to be made... @ras0219 @ras0219-msft @strega-nil @strega-nil-ms @vicroms @dan-shaw do you have any ideas?

**Reply by strega-nil-ms** [reply]
I think this would be pretty easy, assuming that `-Wl,-rpath,'$ORIGIN'` doesn't require any special tools besides a standard linker.

**Reply by Neumann-A** [reply]
add a new toolchain (like https://github.com/microsoft/vcpkg/blob/master/scripts/toolchains/linux.cmake) call it `linux-dynamic.cmake` (or do magic flags via inspection of `VCPKG_LIBRARY_LINKAGE` as the windows toolchain does) and add the linker flag there. 
Maybe decide on a better location for toolchains then `script/toolchains`, e.g. `triplet/toolchains` (since the toolchain is really part of the triplet) and `set(VCPKG_CHAINLOAD_TOOLCHAIN_FILE "${CMAKE_CURRENT_LIST_DIR}/toolchains/linux-dynamic.cmake")`

**Reply by dg0yt** [reply]
I would prefer an rpath of `$ORIGIN/../lib`, and post-build adjustments as needed for vcpkg's unusual file system layout for executables. Or the combination `$ORIGIN/../lib:$ORIGIN` which would allow for traditional UNIX layouts and for Windows-style side-by-side deployment.

**Reply by Osyotr** [reply]
@dg0yt this would also fix ports that don't copy tools dependencies.

**Reply by Neumann-A** [reply]
> I would prefer an rpath of $ORIGIN/../lib, and post-build adjustments as needed for vcpkg's unusual file system layout for executables. Or the combination $ORIGIN/../lib:$ORIGIN which would allow for traditional UNIX layouts and for Windows-style side-by-side deployment.

I would argue that this is a custom triplet/toolchain.

**Reply by dg0yt** [reply]
>  this would also fix ports that don't copy tools dependencies.

@Osyotr It is not so easy. vcpkg doesn't install tools to `bin` or `tools` but to `tools/${PORT}` and `tools/${PORT}/bin`, plus `tools/${PORT}/debug/bin`. Cf. https://github.com/microsoft/vcpkg/issues/17607.
If there was some interest in cleaning up the installation layout for tools, it would be possible to use `$ORIGIN/../../debug/lib:$ORIGIN` for the debug variant. (Copying tool dependencies is mandatory for windows anyway. However, it would be attractive to omit copying on platforms that have rpath.)

**Comment by MrBolt2005** [comment]
Hummm...🤔

Enviado do Email<https://go.microsoft.com/fwlink/?LinkId=550986> para Windows

De: ***@***.***>
Enviado:domingo, 29 de agosto de 2021 13:07
Para: ***@***.***>
***@***.***>; ***@***.***>
Assunto: Re: [microsoft/vcpkg] Make Linux dynamic linkage a first class citizen (#19127)


@MrBolt2005<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FMrBolt2005&data=04%7C01%7C%7C8c5e030d293541c75d1708d96b0727ba%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637658500742941343%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=Er64CIE1SrW7LYcFAT7AA7PySR9g4bC17wA8RCjb8ls%3D&reserved=0> because they use meson, which is an "extraordinary
special kind of flower". There is a bug somewhere when you pass strings with quotes and I really don't want to dig in it.

**Comment by davidchisnall** [comment]
As the POC shows, you can use `-rpath` to define the shared library location.  Note that this can be relative to `$ORIGIN`, which is resolved at load time to be the path of the binary, so you can do exactly the same thing that you do on Windows: install a binary into a directory and install its libraries along with it.  It's generally considered bad form to put libraries in a `bin` directory on *NIX platforms and so the common solution for software wanting to use this model is to set rpath to something like `$ORIGIN/../lib/myprogname/` and install shared libraries in `$PREFIX/lib/myprogname`.

One other possibility would be to install shared libraries as `{libname}-{build hash}.so` and expose that as the library name.  This would allow many different versions and configurations of shared libraries to coexist on the same system.

With respect to the original motivation, note that the LGPL does not 'forbid static linking.' and it is categorically not true that 'Using static linking means your code will be GPL licensed.'  I am not a lawyer and neither is the original poster, please talk to a lawyer if you need advice about software licensing.

**Reply by dweckmann** [reply]
You are indeed right, I should have said "If you use static linking, you must release your code as LGPL or provide a way to relink your code with another version of the LGPL library". It is however not so easy to provide a way to relink statically linked commercial proprietary binaries.

Guide for beginner:
https://copyleft.org/guide/

**Comment by Cazadorro** [comment]
We are also running into this issue.  We primarily develop on linux, and while we appreciate the initial decision to mainly use static linking and agree with the reasoning behind it, LGPL and GPL really screw us over on Linux, is there not even a workaround for these issues? We've had to basically halt all VCPKG work in our group because of this, and try to shift gears to Conan because this is such a big legal issue for us.

**Reply by BillyONeal** [reply]
Does conan actually fix this issue? The issue is still pending discussion because the behavior of the OS loader is hostile to a good experience here, not because the vcpkg maintainers don't care.

**Reply by Osyotr** [reply]
UPD: x64-linux-dynamic is now included in vcpkg! https://github.com/microsoft/vcpkg/pull/23035
I've built my app with patches from first PR and it looks promising. No LD_LIBRARY_PATH required, even Qt did successfully found its plugins. The only issue I didn't fix yet is rpath in debug tools but who use them anyway :)? Second PR is also relevant, but I didn't tried it yet.
https://github.com/microsoft/vcpkg/pull/23035
https://github.com/microsoft/vcpkg/pull/19940

**Reply by dweckmann** [reply]
Well our experience with conan was even worse, but things may have changed (we tried it two years ago). The dynamic linking is well supported (really...), BUT, the main issue we had is that not everything are glued together. It's a collection of recipes, some may work with others, some simply not because the version doesn't match, patches break compatibility with third party libraries... Having a project with a lot of dependencies, with custom build option like cuda/cudnn, with openCV, Qt, VTK, was just very hard to maintain. As I said, maybe things are in a better shape now..

Anyway, for us, it was finally too complicated and time consuming, so we decided to stick to default system libraries and to stick to a particular Linux distribution (the last Ubuntu), changing libraries when upgrading the distribution. Retrospectively, the sanest choice we did. 

We really would love to have a working dynamic VCPKG on Linux, it would allow us to share the same version / build system, and to use it on any linux distribution easily. The dynamic shared library support is is really the thing that have stopped us to use VCPKG on Linux.

My two cents...

**Reply by Cazadorro** [reply]
@BillyONeal Conan works as you'd expect, with the problems you'd expect AFAIK, you use RPATH.  We basically don't want to use the system libraries *ever* on linux, because we run into so many outdated libs from the system repo, so same process dynlib names causing issues shouldn't be a problem for us.  Forcing RPATH works for us.  

* If we didn't have to worry about these licensing issues, we'd be 100% happy with the solution presented by VCPKG.  Licensing really is the sticking point here. 
* We still *want* static linking for virtually everything else that isn't LGPL. We don't need a general purpose solution that works for every library.

We also are just looking for a work around, that would be enough for us.

**Reply by JackBoosY** [reply]
The most suitable solution needs to be discussed in depth.

**Comment by RealLitb** [comment]
I haven't yet understood what exactly the problem is. Is it that Linux _also_ searches its system provided libraries? Link the programs with `-z nodeflib`.  Is it that Linux finds system provided libraries first? Link with `-rpath` and make use of `$ORIGIN`. 

Is it that Linux shared libraries have a single global namespace for all globally visible symbols? That's by design. Why would that be a problem if you take the first two measures?

I'm not experienced with VCPKG, but why do you want to forbid loading of system libraries? If VCPKG is building a self-contained system, then it sounds to me the correct choice would be to execute the binaries in its own container/chroot instead.

**Reply by BillyONeal** [reply]
It isn't a problem if some system libraries get picked, but it is a problem if system libraries get picked over ones vcpkg explicitly built. Like if you built libcurl 1.78 in vcpkg but the system has 1.60, that developer expects to get curl 1.78, not 1.60, and will likely experience silent ABI break hell as a result.

Moreover, when I first wrote my response in 2020 we had absolutely no way to get the vcpkg libraries to be found *ever*, it was only system ones. As a result, deploying dynamic bits on *nix was not possible, so we wouldn't even add a community triplet targeting linux-dynamic.

> Link with -rpath and make use of $ORIGIN

Most build systems do not have plumbing to hook this up. Someone did recently-ish contribute stuff that uses patchelf to set reasonable rpath values that can work.

**Reply by RealLitb** [reply]
I think I understand the issue now. Say there is a libboost.so in the system but a program wants its own libboost for not putting that burden on the user. So the program should use the VCPKG's libboost, but may and often must use other libs from the system. The issue is that those other libs may link against libboost too. And probably don't care whether they link against the system's or the VCPKG one. But the program cares and wants the VCPKG one. 

I'm not an expert, therefore I do not yet see the potential for real memory corruption. My informed beginner opinion is this: If two SO files are incompatible, they should not get the same SOname. So if the system libraries _do_ care that they do get their own libboost instead of third-party provided ones, the upstream libboost authors possibly (to stay with the example) should have had that incompatibility in mind and give their library different SOnames because apparently their libs have ABI-breaking configurations.

Maybe VCPKG should add a "vcpkg" into the SONAME of each library and use an ldconfig  pass afterwards to create symlinks named according to the SO-name (the non-link on-disk name doesn't need to be different).

**Reply by BillyONeal** [reply]
> If two SO files are incompatible, they should not get the same SOname

The vast Vast VAST majority of the C++ world does not care about ABI, and does not follow that. You deploy the program hermetically with its dependencies, or you're toast.

This is also why binary deployment ever on *nix is such a pain and why users are often forced into using their particular distro feed's copy of libraries.

> Maybe VCPKG should add a "vcpkg" into the SONAME of each library and use an ldconfig pass afterwards to create symlinks named according to the SO-name (the non-link on-disk name doesn't need to be different).

That would cause basically all dependency finding to fail, since build systems aren't going to be looking for our funny renamed copies.

It's just a huge mess to do dynamic linking on *nix unless you have absolute systemwide control like apt/dnf/etc. do. The blessing is that you don't have every application on the system deploying their own universe of lib dependencies, the curse is that applications are kind of forced to play ball with the system's supplied package management system.

**Reply by BillyONeal** [reply]
> If two SO files are incompatible, they should not get the same SOname

Also this is no help because the SOname is gone once you get to process symbol table time. Only one boost::Xxx gets picked.

**Reply by RealLitb** [reply]
> > Maybe VCPKG should add a "vcpkg" into the SONAME of each library and use an ldconfig pass afterwards to create symlinks named according to the SO-name (the non-link on-disk name doesn't need to be different).
> 
> That would cause basically all dependency finding to fail, since build systems aren't going to be looking for our funny renamed copies.
> 

The trick is that the copy is not renamed. The file would still be called `libboost.so.1.2.3` but if you run ldconfig on it, a symlink `liboost-vcpkg.so.1.2.3` is created and points to it. At least that was my trick-plan. Not sure if it would work out. But I didn't think about the runtime symbol table. As you said correctly, only one symbol of them gets picked.

**Reply by Osyotr** [reply]
FYI, `x64-linux-dynamic` is usable now. IIRC, the reason why it is not tested in CI is because there's no budget for it.

**Reply by RealLitb** [reply]
I tested my trick  but to prevent symbol collision at runtime, I also passed `--default-symver`, which automatically adds the SOname to the dynamic symbol name. I have no idea what packages this approach would break. Given this source:

    #if !MAIN
    void shouldbeversioned() { }
    #else
    void shouldbeversioned();
    int main() {
        shouldbeversioned();
    }
    #endif

And this command line

    g++ -std=c++20 -O2 -Wall -pedantic -pthread -shared -fPIC -Wl,--default-symver -Wl,-soname libaout-vcpkg.so -o libaout.so main.cpp && readelf -W --dyn-syms -d libaout.so && g++ -std=c++20 -L. -laout -O2 -Wall -pedantic -pthread -DMAIN=1 main.cpp && ln -s libaout.so libaout-vcpkg.so && readelf -W --dyn-syms -d a.out && LD_LIBRARY_PATH=. ./a.out && echo success

It outputs


```
Dynamic section at offset 0x680 contains 28 entries:

  Tag        Type                         Name/Value

 0x0000000000000001 (NEEDED)             Shared library: [libstdc++.so.6]

 0x0000000000000001 (NEEDED)             Shared library: [libm.so.6]

 0x0000000000000001 (NEEDED)             Shared library: [libgcc_s.so.1]

 0x0000000000000001 (NEEDED)             Shared library: [libpthread.so.0]

 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]

 0x000000000000000e (SONAME)             Library soname: [libaout-vcpkg.so]

....


Symbol table '.dynsym' contains 13 entries:

   Num:    Value          Size Type    Bind   Vis      Ndx Name

....
     6: 00000000000005e0     1 FUNC    GLOBAL DEFAULT   11 _Z17shouldbeversionedv@@libaout-vcpkg.so
...

Dynamic section at offset 0x7b8 contains 29 entries:

  Tag        Type                         Name/Value

 0x0000000000000001 (NEEDED)             Shared library: [libaout-vcpkg.so]

 0x0000000000000001 (NEEDED)             Shared library: [libstdc++.so.6]

 0x0000000000000001 (NEEDED)             Shared library: [libm.so.6]

 0x0000000000000001 (NEEDED)             Shared library: [libgcc_s.so.1]

 0x0000000000000001 (NEEDED)             Shared library: [libpthread.so.0]

 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
.....

Symbol table '.dynsym' contains 8 entries:

   Num:    Value          Size Type    Bind   Vis      Ndx Name

     0: 0000000000000000     0 NOTYPE  LOCAL  DEFAULT  UND 

     1: 0000000000400674     0 FUNC    GLOBAL DEFAULT   14 _fini

     2: 00000000004004b0     0 FUNC    GLOBAL DEFAULT   10 _init

     3: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND _Z17shouldbeversionedv@libaout-vcpkg.so (2)

     4: 0000000000000000     0 NOTYPE  WEAK   DEFAULT  UND _ITM_deregisterTMCloneTable

     5: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND __libc_start_main@GLIBC_2.2.5 (3)

     6: 0000000000000000     0 NOTYPE  WEAK   DEFAULT  UND __gmon_start__

     7: 0000000000000000     0 NOTYPE  WEAK   DEFAULT  UND _ITM_registerTMCloneTable

success


```

It demonstrates that you can link to `-laout` (because there's a `libaout.so`, which is usually a (-dev) symlink aswell) but then at runtime the `aout-vcpkg` is loaded and also the symbols are versioned with the `aout-vcpkg` in it. I'm not experienced with versioning, but maybe this or a variant of it could give a rock-solid experience.

**Reply by RealLitb** [reply]
> FYI, `x64-linux-dynamic` is usable now. IIRC, the reason why it is not tested in CI is because there's no budget for it.

Like @BillyONeal said (and I think he has a good point), the major reason seems to be that system-SO files will correctly load a required system lib from the system but as symbols are same-named, the dynamic symbol resolution resolves symbols in them in favour of VCPKG's versions (they are first in breath-first order). This issue is not fixed by -rpath.

**Comment by VioletGiraffe** [comment]
There is now `x64-linux-dynamic`, thank you for that, but why no `arm64-linux-dynamic`?

**Comment by MaxRev-Dev** [comment]
@BillyONeal @dg0yt What if the `rpath` gets overridden? Could you please give an advice here?

Consider a .NET app which has `libssl.so.3` dependency running on Debian 12 ([SSL 3 breaking change](https://github.com/dotnet/docs/issues/35953)). This `libssl.so.3` has version `3.0.2`. Docs on how dotnet loads native libraries [are here](https://github.com/dotnet/core/blob/main/Documentation/self-contained-linux-apps.md).

So, I'm building `curl` with `openssl` dependency of version 3.2.0.
On start the app loads the system `ibssl.so.3` (3.0.2). Then it fails on curl.
```
/lib/x86_64-linux-gnu/libssl.so.3: version `OPENSSL_3.2.0' not found (required by /app/bin/Debug/net8.0/runtimes/linux-x64/native/libcurl.so.4)
```

**Reply by MaxRev-Dev** [reply]
Should the SSL version be more strict in this case including the `minor` and `fix` versions in naming? This should be hooked up by linker

**Reply by Osyotr** [reply]
Provide an empty overlay for openssl?

**Reply by dg0yt** [reply]
The short question probably isn't a good answer... 
1. Your final application needs to be made aware of the runtim lib locations, too. Did you set it up at all?
2. Basically you shouldn't mix different binaries of the same library. If you are forced to use the system openssl library, **you can use an empty overlay ports for openssl** to make all binaries in vcpkg use that system library. But this quickly becomes an advanced task.

**Reply by MaxRev-Dev** [reply]
@dg0yt This one is already a bit complicated, but anyways..
 1.  It looks like I'm forced to use SSL loaded by runtime. I'm compiling packages for a [library](https://github.com/MaxRev-Dev/gdal.netcore/actions/runs/7545147163/job/20541620227#step:6:9481). The runpath/rpath is set to the following.
`  0x000000000000001d (RUNPATH)            Library runpath: [$ORIGIN:$ORIGIN/runtimes/linux-arm64/native]`
I tried to set LD_LIBRARY_PATH to point to the executable directory for a single run and it **works**. However, it's not a workaround at all as I can't include that into the library without an additional executable wrapper. 
But regarding loader paths I'm getting the following LD_DEBUG output:

<details><summary>Details</summary>
<p>

```shell
     15675:
     15675:     calling init: /usr/share/dotnet/shared/Microsoft.NETCore.App/8.0.0/libSystem.Security.Cryptography.Native.OpenSsl.so
     15675:
     15675:     find library=libssl.so.3 [0]; searching
     15675:      search cache=/etc/ld.so.cache
     15675:       trying file=/lib/aarch64-linux-gnu/libssl.so.3
     15675:
     15675:     find library=libcrypto.so.3 [0]; searching
     15675:      search cache=/etc/ld.so.cache
     15675:       trying file=/lib/aarch64-linux-gnu/libcrypto.so.3
     15675:
     15675:
     15675:     calling init: /lib/aarch64-linux-gnu/libcrypto.so.3
     15675:
     15675:
     15675:     calling init: /lib/aarch64-linux-gnu/libssl.so.3
     15675:
     15675:     /lib/aarch64-linux-gnu/libssl.so.3: error: symbol lookup error: undefined symbol: SSL_state (fatal)
     15675:     /lib/aarch64-linux-gnu/libssl.so.3: error: symbol lookup error: undefined symbol: ERR_put_error (fatal)
     15675:     find library=libc.so [0]; searching
     15675:      search cache=/etc/ld.so.cache
     15675:      search path=/lib/aarch64-linux-gnu:/usr/lib/aarch64-linux-gnu:/lib:/usr/lib            (system search path)
``` 

</p>
</details> 


2. It should be possible to inject an empty overlay into the build pipeline as I have several dependencies on OpenSSL.  
However, I'm building on Debian 11 as it has the lowest widely used version of glibc. It has libssl1.1 from the system packages. I moved from CentOS as it was difficult to maintain all dependencies. Then I started using Debian 11 until last year when the dotnet team moved to Debian 12. I'm not sure to which version stick to.

I saw [this comment](https://github.com/microsoft/vcpkg/discussions/19127#discussioncomment-1049622) by @BillyONeal  on linking statically, but that would increase package size.  And for now, it's not an option because GDAL fails to link to a static `curl` library. I'd consider it as a last resort choice.

For now, I tried to specify the major/minor versions and it works without any issues. 
Can I include that as a post-build step but would that introduce any issues? 
And the question remains - is it possible to extend naming to include specific versions for shared libraries with vcpkg?


```
patchelf --replace-needed libssl.so.3 libssl.so.3.2.0 libcurl.so.4
patchelf --replace-needed libcrypto.so.3 libcrypto.so.3.2.0 libssl.so.3
mv libcrypto.so.3 libcrypto.so.3.2.0
mv libssl.so.3 libssl.so.3.2.0
```

**Reply by dg0yt** [reply]
> And for now, it's not an option because GDAL fails to link to a static curl library.

Not a known issue. But with curl, we quickly get to openssl again.

I can't answer the other questions for this context.

**Reply by MaxRev-Dev** [reply]
@dg0yt thank you

**Reply by MaxRev-Dev** [reply]
Regarding SSL [version compatibility](https://www.openssl.org/policies/releasestrat.html#:~:text=API/ABI%20compatibility%20will%20be%20guaranteed%20for%20the%20same%20MAJOR%20version%20number.), it looks like I can do this patch by specifying the minor/patch version.
Could anyone from the VCPKG team advise on how to add the patch version to the package?
Thank you

**Comment by tsondergaard** [comment]
Not having dynamic linking supported as a first-class citizen on Linux is a problem for my company's project as well. One of our products alone consists of 30+ executables. Many of them use Qt libraries which are large libraries so this would make our installation many times larger. We also use Qt Webengine which cannot be built as a static library. Finally we use Qt under the LGPL license, which makes static linking a problem.

@BillyONeal wrote:
> We choose static linking on the POSIX-ish platforms because dynamic linking generally forces you to use the system-provided libraries (e.g. those acquired through apt). It doesn't matter if you want to use the vcpkg version of boost in general, for example, because all the operating system's default load library paths are going to look in places owned by apt or dnf or similar.

With the proper RUNPATH set this is a manageable problem.

Yes, there are caveats and dragons with dynamic linking but as I have pointed out above it is simply required for a lot of projects.
